DEVPROD-7418 optimize hostsCanRunTasksQuery for task queue page

khelif96 opened the pull request:
DEVPROD-7418

### Description
The task queue page performs a query to fetch the number of hosts in the task queue for every distro. Sometimes this query becomes extremely poorly performing and can take > 100s to complete. My initial hypothesis was that this was due to the shape of the query causing it to sometimes use an incorrect index. In a previous [pr](https://github.com/evergreen-ci/evergreen/pull/7890) I introduced a hint to guide the database to using the correct index. 
It quickly became apparent that this didn't work. Even though the query always used the same index it exhibited dramatically different performance. In this [example](https://ui.honeycomb.io/mongodb-4b/environments/production/datasets/evergreen/result/t8Xj4nXw5GC/trace/BTVktyjsjpW?fields%5B%5D=s_name&fields%5B%5D=s_serviceName&span=50564a792f3775bf) out of 770 aggregations 10 of them took between 60-200s to execute. 
After consulting with TSE's it turns out the issue wasn't the index but was instead the query planner. In some cases (<1% of the time) the query planner would choose the wrong plan to execute the query and choose an extremely inefficient plan that would not actually use the status portion of the index causing the db to perform a fetch on millions of documents. 
The _good_ plan and _bad_ plan were both assigned the same score by the query planner and sometimes the query planner would choose the slow _bad_ plan instead of the correct one.
Updating the query shape helps improve the query planner's ability to choose the correct plan.  

This is a known performance issue in MongoDB 7.0 and should be [improved](https://jira.mongodb.org/browse/SPM-3316) in newer versions of the DB. 

While the new query is not the same as the previous query and may return slightly different results I think the tradeoff is acceptable since the performance of the page atm makes it unusable. 

### Testing
Tested this locally and generated 2 explain plans before and after showing a difference 

[explain before](https://github.com/user-attachments/files/15892254/output.1.json)
[explain after](https://github.com/user-attachments/files/15892229/output.json)

khelif96 commented:
I will update the tests if people agree with this approach of having less correct host counts.

ablack12 reviewed (approved):
Can we file a ticket to change this back after we've upgraded?

khelif96 commented:
> Can we file a ticket to change this back after we've upgraded?

I changed the approach for this and instead used a `$or` that wraps both conditions. This approach ensures consistency while also causing the query planner to only come up with 1 plan eliminating the risk of choosing the wrong plan.

Updated explain plan
[output.json](https://github.com/user-attachments/files/15929970/output.json)

ablack12 commented on model/host/db.go:
```diff
@@ -239,19 +239,23 @@ func runningHostsQuery(distroID string) bson.M {
 		"$or": []bson.M{
 			{
-				StatusKey: evergreen.HostRunning,
+				distroIDKey:  distroID,
```
I'd add a comment here to explain why you did this, so a well-intentioned engineer doesn't come by and try to simplify things

khelif96 replied:
Done!

ablack12 reviewed (commented):
Works for me, can we check on this after deploy before closing the ticket to make sure this is good?

khelif96 commented:
> Can we file a ticket to change this back after we've upgraded?

Ticket to revert: https://jira.mongodb.org/browse/DEVPROD-8360
